rules.txt

```diff
@@ -0,0 +1,40 @@
+rules_version = '2';
+service cloud.firestore {
+  match /databases/{database}/documents {
+    function signedIn() {
+      return request.auth.uid != null && request.auth.uid != null;
+    }
+    
+    function isDocOwner() {
+      return request.auth.uid == request.resource.data.owner;
+    }
+    
+    function updatedDocHasCorrectOwner() {
+      return request.auth.uid == request.resource.data.owner;
+    }
+    
+    function updatedDocHasSameOwner() {
+      return resource.data.owner == request.resource.data.owner;
+    }
+    
+    function updatedDocHasSameSharing() {
+      return resource.data.sharedWith == request.resource.data.sharedWith;
+    }
+    
+    function isSharedWithMe() {
+      return request.auth.token.email in resource.data.sharedWith;
+    }
+    
+    match /App_Data/{document} {
+      allow read: if signedIn() && isSharedWithMe();
+      allow create: if signedIn() || updatedDocHasCorrectOwner();
```

Comment:
Should be &&. This allows any signed-in user to create a document with any owner.

You need more &&'s in next few lines, too.